### PR TITLE
FALCON-2184 Support to delete a non trusted recipe metadata from server

### PR DESCRIPTION
**Pull request by PraveenAdlakha**

### CODE REVIEW

**Comment by PraveenAdlakha** [comment]
Dev Testing : 
dataqa@blr-test-129:/usr/local/falcon/falcon-0.11-SNAPSHOT$ ./bin/falcon extension  -unregister -extensionName hdfs-mirrorings
Extension:hdfs-mirroringsis not registered with falcon.

dataqa@blr-test-129:/usr/local/falcon/falcon-0.11-SNAPSHOT$ ./bin/falcon extension  -unregister -extensionName hdfs-mirroring
ERROR: Internal Server Error;java.lang.Exception: hdfs-mirroringis trusted cannot be deleted.
CausedBy: hdfs-mirroringis trusted cannot be deleted.

dataqa@blr-test-129:/usr/local/falcon/falcon-0.11-SNAPSHOT$ ./bin/falcon extension  -unregister -extensionName test
Deleted entry for:test

File: common/src/main/java/org/apache/falcon/persistence/PersistenceConstants.java [hunk 1]
```diff
@@ -70,4 +70,7 @@ private PersistenceConstants(){
     public static final String GET_ALL_EXTENSIONS = "GET_ALL_EXTENSIONS";
     public static final String GET_EXTENSION_LOCATION = "GET_EXTENSION_LOCATION";
     public static final String DELETE_EXTENSIONS_OF_TYPE = "DELETE_EXTENSIONS_OF_TYPE";
+    public static final String DELETE_ALL_TRUSTED_EXTENSIONS = "DELETE_ALL_TRUSTED_EXTENSIONS";
```
Comment (pallavi-rao) [inline common/src/main/java/org/apache/falcon/persistence/PersistenceConstants.java]: Not constant is not required.

File: extensions/src/main/java/org/apache/falcon/extensions/store/ExtensionStore.java [hunk 2]
```diff
@@ -250,6 +250,20 @@ public String getExtensionResource(final String resourcePath) throws StoreAccess
         return extesnionList;
     }
 
+    public String deleteExtensionMetadata(final String extensionName) throws StoreAccessException{
+        ExtensionType extensionType = AbstractExtension.isExtensionTrusted(extensionName) ? ExtensionType.TRUSTED
+                : ExtensionType.CUSTOM;
+        if (extensionType.toString().equalsIgnoreCase(ExtensionType.TRUSTED.toString())){
+            throw new StoreAccessException(new Exception(extensionName + " is trusted cannot be deleted."));
```
Comment (pallavi-rao) [inline extensions/src/main/java/org/apache/falcon/extensions/store/ExtensionStore.java]: Nit : No need to create new Exception. You can directly create StoreAccessException

File: extensions/src/main/java/org/apache/falcon/extensions/store/ExtensionStore.java [hunk 3]
```diff
@@ -250,6 +250,20 @@ public String getExtensionResource(final String resourcePath) throws StoreAccess
...
+        }
+        if (metaStore.checkIfExtensionExists(extensionName)) {
+            metaStore.deleteExtensionMetadata(extensionName);
+            return "Deleted entry for:" + extensionName;
```
Comment (pallavi-rao) [inline extensions/src/main/java/org/apache/falcon/extensions/store/ExtensionStore.java]: Nit : Use "Deleted extension " instead.

File: extensions/src/main/java/org/apache/falcon/extensions/store/ExtensionStore.java [hunk 4]
```diff
@@ -250,6 +250,20 @@ public String getExtensionResource(final String resourcePath) throws StoreAccess
...
+        }else {
+            return "Extension:" + extensionName + " is not registered with falcon.";
```
Comment (pallavi-rao) [inline extensions/src/main/java/org/apache/falcon/extensions/store/ExtensionStore.java]: Nit : Capitalize F for Falcon.

File: extensions/src/main/java/org/apache/falcon/extensions/store/ExtensionStore.java [hunk 5]
```diff
@@ -250,6 +250,20 @@ public String getExtensionResource(final String resourcePath) throws StoreAccess
         return extesnionList;
     }
 
+    public String deleteExtensionMetadata(final String extensionName) throws StoreAccessException{
+        ExtensionType extensionType = AbstractExtension.isExtensionTrusted(extensionName) ? ExtensionType.TRUSTED
+                : ExtensionType.CUSTOM;
+        if (extensionType.toString().equalsIgnoreCase(ExtensionType.TRUSTED.toString())){
```
Comment (pallavi-rao) [inline extensions/src/main/java/org/apache/falcon/extensions/store/ExtensionStore.java]: You can directly do an enum comparison. No need to convert to string.

extensionType.equals(ExtensionType.TRUSTED)

File: extensions/src/main/java/org/apache/falcon/extensions/store/ExtensionStore.java (hunk 2 above)
Comment (PraveenAdlakha) [inline extensions/src/main/java/org/apache/falcon/extensions/store/ExtensionStore.java]: These are the only two constructors available : 

 public StoreAccessException(String message, Exception e) {
        super(message, e);
    }

    public StoreAccessException(Exception e) {
        super(e);
    }

Comment (pallavi-rao) [inline extensions/src/main/java/org/apache/falcon/extensions/store/ExtensionStore.java]: Can add a method to StoreAccessException (it is part of Falcon codebase). 

Anyway, in this particular case, wouldn't ValidationException be more suitable?

**Pull request closed without merging**